javascript-algorithms-and-data-structures/Learn-Intermediate-OOP-by-Building-a-Platformer-Game/script.js

```diff
@@ -0,0 +1,230 @@
+  { x: 500, y: 450 },
+  { x: 700, y: 400 },
+  { x: 850, y: 350 },
+  { x: 900, y: 350 },
+  { x: 1050, y: 150 },
+  { x: 2500, y: 450 },
+  { x: 2900, y: 400 },
+  { x: 3150, y: 350 },
+  { x: 3900, y: 450 },
+  { x: 4200, y: 400 },
+  { x: 4400, y: 200 },
+  { x: 4700, y: 150 },
+];
+
+const platforms = platformPositions.map(
+  (platform) => new Platform(platform.x, platform.y)
+);
+
+const checkpointPositions = [
+  { x: 1170, y: 80 },
+  { x: 2900, y: 330 },
+  { x: 4800, y: 80 },
+];
+const checkpoints = checkpointPositions.map(
+  (checkpoint) => new CheckPoint(checkpoint.x, checkpoint.y)
+);
+
+const animate = () => {
+  requestAnimationFrame(animate);
+
+  ctx.clearRect(0, 0, canvas.width, canvas.height);
+
+  platforms.forEach((platform) => {
+    platform.draw();
+  });
+
+  checkpoints.forEach((checkpoint) => {
+    checkpoint.draw();
+  });
+
+  player.update();
+
+  if (keys.rightKey.pressed && player.position.x < 400) {
+    player.velocity.x = 5;
+  } else if (keys.leftKey.pressed && player.position.x > 100) {
+    player.velocity.x = -5;
+  } else {
+    player.velocity.x = 0;
+
+    if (keys.rightKey.pressed) {
+      platforms.forEach((platform) => {
+        platform.position.x -= 5;
+      });
+      checkpoints.forEach((checkpoint) => {
+        checkpoint.position.x -= 5;
+      });
+    } else if (keys.leftKey.pressed) {
+      platforms.forEach((platform) => {
+        platform.position.x += 5;
+      });
+      checkpoints.forEach((checkpoint) => {
+        checkpoint.position.x += 5;
+      });
+    }
+  }
+
+  platforms.forEach((platform) => {
+    const collisionDetectionRules = [
+      player.position.y + player.height <= platform.position.y,
+      player.position.y + player.height + player.velocity.y >=
+        platform.position.y,
+      player.position.x >= platform.position.x - player.width / 2,
+      player.position.x <=
+        platform.position.x + platform.width - player.width / 3,
+    ];
+
+    if (collisionDetectionRules.every((rule) => rule)) {
+      player.velocity.y = 0;
+    }
+  });
+
+  checkpoints.forEach((checkpoint, index) => {
+    if (
+      player.position.x >= checkpoint.position.x &&
+      activeCheckpointCollisionDetection
+    ) {
+      if (index === checkpoints.length - 1) {
+        showEndingTitleScreen();
+        activeCheckpointCollisionDetection = false;
+        movePlayer("ArrowRight", 0, false);
+      }
+    }
+  });
+};
+
+const keys = {
+  rightKey: {
+    pressed: false,
+  },
+  leftKey: {
+    pressed: false,
+  },
+};
+
+const movePlayer = (key, xVelocity, isPressed) => {
+  switch (key) {
+    case "ArrowLeft":
+      keys.leftKey.pressed = isPressed;
+      if (xVelocity === 0) {
+        player.velocity.x = xVelocity;
+      }
+      player.velocity.x -= xVelocity;
+      break;
+
+    case "ArrowUp":
+    // older browsers might use spacebar instead of an empty string
+    case " " || "Spacebar":
+      player.velocity.y -= 8;
+      break;
```

Comment:
```suggestion
    case "ArrowUp":
    // older browsers might use spacebar instead of an empty string
    case " ":
    case "Spacebar":
      player.velocity.y -= 8;
      break;
```

Nitpicking, but we shouldn't use conditional operators in a case condition